[v3] Late service registration and error handling overhaul

fbbdev opened the pull request:
# Description

This PR provides three QOL improvements for the runtime side of the binding/service system and fixes a major bug:

- it adds a method `App.RegisterService` for registering services after app creation (but before `App.Run`);
- it adds rich error handling for binding calls inspired by the v2 approach, with json marshaling as default behaviour and global + per-service `MarshalError` functions for custom handling (see the updated docs);
- it standardises and documents the order in which services are started and shut down (registration order and reverse registration order, respectively);
- it fixes a macOS bug whereby service shutdown methods would not be called on quit (due to the fact that on macOS the app event loop never returns control back to `App.Run`).

Additionally, it adds a test harness for end-to-end tests of the application object and uses it to test service startup/shutdown sequences and error handling. Note that the go test framework spawns one process per package, hence – against go conventions – I had to split tests over many packages because at most one application run per process is supported.

### 🚨 Breaking changes 🚨

There are a few minor breaking changes necessitated by the restructuring described above (should not affect users in practice):

- `ServiceStartup` hooks are now invoked when `App.Run` is called, not in `application.New`
- `ServiceStartup` errors are now returned from `App.Run` instead of terminating the process
- Binding and dialog calls from JS now reject with rich error objects instead of simple strings

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Ran the test suite including new and updated tests, verified that everything is passing; tested examples manually.

- [ ] Windows
- [x] macOS
- [ ] Linux

## Test Configuration

```
 Wails (v3.0.0-dev)  Wails Doctor

# System

┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
| Name          | MacOS                                                                                                          |
| Version       | 12.6.6                                                                                                         |
| ID            | 21G646                                                                                                         |
| Branding      | Monterey                                                                                                       |
| Platform      | darwin                                                                                                         |
| Architecture  | amd64                                                                                                          |
| Apple Silicon | unknown                                                                                                        |
| CPU           | Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz                                                                       |
| CPU           | Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz                                                                       |
| GPU           |  cores, Metal Family: Supported, Metal GPUFamily macOS 2                                                       |
|       Metal Family: Supported, Metal GPUFamily macOS 2                                                                         |
| Memory        | 32 GB                                                                                                          |
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

# Build Environment

┌─────────────────────────────────────────────────────────┐
| Wails CLI    | v3.0.0-dev                               |
| Go Version   | go1.23.5                                 |
| Revision     | 7f6b911e15fccadd140077b809d244a9878bf2c6 |
| Modified     | true                                     |
| -buildmode   | exe                                      |
| -compiler    | gc                                       |
| CGO_CFLAGS   |                                          |
| CGO_CPPFLAGS |                                          |
| CGO_CXXFLAGS |                                          |
| CGO_ENABLED  | 1                                        |
| CGO_LDFLAGS  |                                          |
| GOAMD64      | v1                                       |
| GOARCH       | amd64                                    |
| GOOS         | darwin                                   |
| vcs          | git                                      |
| vcs.modified | true                                     |
| vcs.revision | 7f6b911e15fccadd140077b809d244a9878bf2c6 |
| vcs.time     | 2025-02-12T08:35:30Z                     |
└─────────────────────────────────────────────────────────┘

# Dependencies

┌───────────────────────────┐
| npm             | 10.8.2  |
| *NSIS           | v3.10   |
| Xcode cli tools | 2395    |
└─ * - Optional Dependency ─┘

# Checking for issues

 SUCCESS  No issues found

# Diagnosis

 SUCCESS  Your system is ready for Wails development!
```

# Checklist:

- [x] I have updated `docs/src/content/docs/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a post-shutdown hook to allow custom code execution after shutdown.
  - Added a new error type for improved fatal error detection, along with a streamlined service registration approach.

- **Improvements**
  - Upgraded error handling in JavaScript bindings and dialogs so that errors now return standardized error objects.
  - Enhanced the reliability of service startup and shutdown order for better lifecycle management.
  - Improved error handling and logging consistency across various components.
  - Updated error handling in multiple methods to use structured error messages and improved formatting.
  - Standardized documentation for service lifecycle management and error handling.

- **Documentation**
  - Expanded and clarified content on error handling and service lifecycle procedures.

- **Tests**
  - Added a test harness to validate application startup and shutdown sequences.
  - Introduced comprehensive tests for service startup and shutdown behaviors, including error handling scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fbbdev commented on v3/internal/assetserver/assetserver.go:
```diff
@@ -126,11 +129,8 @@ func (a *AssetServer) serveHTTP(rw http.ResponseWriter, req *http.Request, userH
-		a.services = make(map[string]http.Handler)
-	}
-	a.services[prefix] = handler
+func (a *AssetServer) AttachServiceHandler(route string, handler http.Handler) {
+	a.services = append(a.services, service{route, handler})
 }
```
The old behaviour was to have a map with routes as keys and do a linear scan of the map. This was worse than a slice for two reasons: first, route collisions would result in silent substitutions, and second the iteration order was non-deterministic.

The slice approach is not the best possible, but enforces a clear rule: priority goes to the first in registration order.

We might want to implement an actual trie of path segments here and warn about duplicates @leaanthony
A much quicker and easier solution might be a `http.ServeMux`, but we'll have to work around its redirect behaviour or at least emit errors in case of redirects for the moment being.

leaanthony replied:
I'm ok with this for now 👍

fbbdev commented on v3/pkg/application/messageprocessor_call.go:
```diff
@@ -63,86 +78,116 @@ func (m *MessageProcessor) processCallMethod(method int, rw http.ResponseWriter,
+
 			if result != nil {
 				// convert result to json
 				jsonResult, err = json.Marshal(result)
```
These errors are logged at `processCallMethod`. The error from `json.Marshal` is printed and should provide clues to debug the problem. I think printing the raw result might be overkill, it's probably not going to be readable.

fbbdev commented on v3/internal/runtime/desktop/@wailsio/runtime/src/calls.js:
```diff
@@ -48,22 +48,66 @@ function generateID() {
  *
  * @param {string} id - The id of the promise handler.
- * @param {string} message - The error message to reject the promise handler with.
+ * @param {string} data - The error data to reject the promise handler with.
+ * @param {boolean} isJSON - Indicates whether the data is JSON or not.
  *
```
Promise.resolve should not ever throw (then/catch callbacks are scheduled as microtasks, not run immediately), hence the try/catch block should be fine.

fbbdev commented on v3/pkg/application/messageprocessor_clipboard.go:
```diff
@@ -15,30 +17,31 @@ var clipboardMethods = map[int]string{
 		return
 	}
 
+	m.Info("Runtime call:", "method", "Browser."+browserMethods[method], "text", text)
```
Good catch!

fbbdev commented:
It seems the new tests are failing randomly on ubuntu with a message about not being able to open a display (some of them pass though). Will have to spin up a vm and check tomorrow.

fbbdev commented:
Perhaps adding this to the pipelines

> https://github.com/marketplace/actions/setup-xvfb

might be enough to solve the issue

fbbdev commented:
Updated with
- virtual display for ubuntu testing pipelines
- fixed error formatting everywhere

⚠️ COMMIT `33d0004` CHANGES A GITHUB ACTIONS WORKFLOW, INSPECT CAREFULLY ⚠️ 

Pipeline run in my repo: https://github.com/fbbdev/wails/actions/runs/13355679249

fbbdev commented:
Formatting errors completely fixed.

Ubuntu tests are randomly timing out waiting for some desktop service 😩

Looking for a solution.

fbbdev commented:
All lights green at last 🥳 
https://github.com/fbbdev/wails/actions/runs/13371757132

This is ready to merge.

Had to fix `wails3 init` as well, a change in Go 1.24 broke dev mode detection hence template tests were not passing:

> The go build command now sets the [main module’s version](https://go.dev/pkg/runtime/debug#BuildInfo.Main) in the compiled binary based on the version control system tag and/or commit. A +dirty suffix will be appended if there are uncommitted changes.

https://go.dev/doc/go1.24#go-command
Dev mode detection previously relied on the main module's version being `(devel)`.

fbbdev commented on v3/pkg/application/internal/tests/services/startup/startup_test.go:
```diff
@@ -0,0 +1,102 @@
+	validate(t, services[5], 4, 5)
+}
+
+func validate(t *testing.T, svc application.Service, prev ...int64) {
+	id := svc.Instance().(interface{ Id() int }).Id()
+	seq := svc.Instance().(interface{ StartupSeq() int64 }).StartupSeq()
+
+	if seq == 0 {
+		t.Errorf("Service #%d did not start up", id)
+		return
+	}
+
+	for _, p := range prev {
+		if seq <= p {
+			t.Errorf("Wrong startup sequence number for service #%d: wanted >%d, got %d", id, p, seq)
+		}
+	}
+}
```
@coderabbitai unnecessary, these are mock services that are guaranteed to implement the interfaces

fbbdev commented on v3/internal/commands/update_cli.go:
```diff
@@ -31,9 +32,9 @@ func UpdateCLI(options *UpdateCLIOptions) error {
 		term.Println("This Wails CLI has been installed from source. To update to the latest stable release, run the following commands in the `" + v3Path + "` directory:")
 		term.Println("   - git pull")
-		term.Println("   - go install")
+		term.Println("   - wails3 task install")
```
@coderabbitai Good catch, fixing!

fbbdev commented on v3/pkg/application/application.go:
```diff
@@ -660,25 +683,57 @@ func (a *App) Run() error {
+
+		if s, ok := service.instance.(ServiceShutdown); ok {
+			a.debug("Shutting down service:", "name", getServiceName(service))
+			if err := s.ServiceShutdown(); err != nil {
+				a.error("error shutting down service '%s': %w", getServiceName(service), err)
```
@coderabbitai shutdown has already failed at that point 🙄
